src/main/java/io/pinecone/clients/Pinecone.java

```diff
@@ -21,19 +21,49 @@ public Pinecone(String apiKey) {
         this(apiKey, new OkHttpClient());
     }
 
+    public Pinecone(String apiKey, String sourceTag) {
+        this(new PineconeConfig(apiKey), sourceTag, new OkHttpClient());
+    }
+
     public Pinecone(String apiKey, OkHttpClient okHttpClient) {
-        if (apiKey == null || apiKey.isEmpty()) {
-            throw new PineconeValidationException("The API key is required and must not be empty or null");
-        }
-        config = new PineconeConfig(apiKey);
+        this(new PineconeConfig(apiKey), null, okHttpClient);
+    }
```

Comment:
I did a few things here that I think feel a bit better to reason about in terms of encapsulation, but let me know.

1. Added additional constructor overrides to support passing `PineconeConfig` as well as the generic type arguments. `PineconeConfig` is where `sourceTag` is stored, and is used for `PineconeConnection`. People will be able to import and use these classes directly, so it felt like it made sense to make it flexible.
2. Encapsulated the validation logic inside `PineconeConfig.validate()`. We construct the config then validate it before using it to configure the `ApiClient`.
3. Threading `sourceTag` through to `PineconeConfig` where it's set and validated when set.
4. Added a check for `PINECONE_DEBUG` in environment variables and enables logging for the `ApiClient` if so. This was helpful for validating `user-agent` changes.

Reply:
I think with this number of constructor params/combinations the permutations of constructor overloads will be super confusing for users -- it's an anti-pattern known as "telescoping constructors".

For `sourceTag` in particular, it's an optional param that we'll tell partners to explicitly set, so we *could* do something like:
```java
Pinecone pc = new Pinecone("my-api-key");
pc.setSourceTag("my-tag");
```

A safer approach, especially if there are lots of such optional params/configurations would be to use a builder pattern. It would look something like this:

```java
public class Pinecone {
  private final PineconeConfig config;

  public static class Builder {
     // required params
     private final String apiKey;

     // optional params
     private String sourceTag;

     public Builder(apiKey) {
       this.apiKey = apiKey;
     }

     public Builder withSourceTag(String sourceTag) {
         this.sourceTag = sourceTag;
         return this;
      }

     public Pinecone build() {
         config = .... // construct PineconeConfig from contents of Builder
         return new Pinecone(config);
     }
  }
}

// and to use it:
Pinecone pc = new Pinecone.Builder("my-api-key").withSourceTag("my-tag").build();
```

Reply:
Thank you, this is good to know. I've heard about the builder pattern, but wasn't entirely clear on what that looks like or when it's employed. I think we should look at moving to this for some areas where we have a lot of variations of optional parameters and complex assortments of constructors. 👍

Reply:
We have flexibility with the naming as well, it doesn't *have* to be Builder. It can be based on the semantics we'd like to provide users with. 

So for example, here are some other options, but you and @rohanshah18 can probably come up with better ones:

```java
Pinecone pc = new Pinecone.Connection("my-api-key").withSourceTag("my-tag").connect();
Pinecone pc = new Pinecone.Client("my-api-key").withSourceTag("my-tag").build();

// we don't have to use "with" either, some APIs skip it
Pinecone pc = new Pinecone.Builder("my-api-key")
      .sourceTag("my-tag")
      .okHttpClient(okClient)
      .build();

// and there's another flavor where we use a static function as a constructor, which would internally use a private builder.
Pinecone pc = Pinecone.newConnection("my-api-key")
      .sourceTag("my-tag")
      .okHttpClient(okClient)
      .build();
```

Reply:
We used to offer something like `withMember1()` in deprecated configs, so I have no issue using `with` keyword. I think its widely used and clear for java users on what the intention is.